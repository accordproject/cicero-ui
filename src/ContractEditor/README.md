## Cicero-UI - ContractEditor

### Usage

```
npm install @accordproject/cicero-ui
```

```
import { ContractEditor } from '@accordproject/cicero-ui';

const clausePropsObject = {
    BODY_FONT (string),
    CLAUSE_BACKGROUND (string),
    CLAUSE_BORDER (string),
    CLAUSE_DELETE (string),
    CLAUSE_DELETE_FUNCTION (function),
    HEADER_FONT (string),
}

const editorPropsObject = {
    BUTTON_BACKGROUND_INACTIVE (string),
    BUTTON_BACKGROUND_ACTIVE (string),
    BUTTON_SYMBOL_INACTIVE (string),
    BUTTON_SYMBOL_ACTIVE (string),
    DROPDOWN_COLOR (string),
    TOOLBAR_BACKGROUND (string),
    TOOLTIP_BACKGROUND (string),
    TOOLTIP (string),
    TOOLBAR_SHADOW (string),
    WIDTH (string),
}

function storeLocal(editor) {
    localStorage.setItem('markdown-editor', editor.getMarkdown());
}

ReactDOM.render(<ContractEditor 
    clauseProps={clauseProps(props.removeFromContract)}
    editorProps={editorPropsObject}
    onChange={storeLocal} />,
  document.getElementById('root'));
```

### Props

- `value` : An `object` which is the initial contents of the editor.
- `onChange` : A callback `function` called when the contents of the editor change.
- `lockText` : A `boolean` to lock all non variable text.
- `loadTemplateObject` : A callback `function` to load a template.
- `parseClause` : A callback `function` to parse the contents of a clause.
- `templates` : An `array` which contains template objects with the following keys: `uir`, `name`, `version`, `description`.
- `editorProps` : An optional `object` for the contract editor styling passed down through an object to Markdown Editor, see below.
- `clauseProps` : An `object` for the clauses in the editor which contains a deletion function, see below.


### Specifications

`editorProps`:
You can style the toolbar of this components, as well as the width of the editor. An object may be passed down this component which will then be picked up by our `markdown-editor`, with the following possible CSS inputs as strings:
- `BUTTON_BACKGROUND_INACTIVE`
- `BUTTON_BACKGROUND_ACTIVE`
- `BUTTON_SYMBOL_INACTIVE`
- `BUTTON_SYMBOL_ACTIVE`
- `DROPDOWN_COLOR`
- `TOOLBAR_BACKGROUND`
- `TOOLTIP_BACKGROUND`
- `TOOLTIP`
- `TOOLBAR_SHADOW`
- `WIDTH`

`clauseProps`:
You can style the Clause Components within the `ContractEditor`. An object may be passed down this component with the following possible CSS inputs as strings, as well as a deletion function:
- `BODY_FONT`
- `CLAUSE_BACKGROUND`
- `CLAUSE_BORDER`
- `CLAUSE_DELETE`
- `CLAUSE_DELETE_FUNCTION`
- `HEADER_FONT`