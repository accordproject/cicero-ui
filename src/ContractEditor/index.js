/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { MarkdownEditor } from '@accordproject/markdown-editor';
import List from '@accordproject/markdown-editor/dist/plugins/list';
import ClausePlugin from './plugins/clausePlugin';

/**
 * A rich text contract editor
 */
class ContractEditor extends React.Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.props.plugins.push(List(), ClausePlugin());
  }

  /**
   * Render this React component
   * @return {*} the react component
   */
  render() {
    return (<MarkdownEditor markdown={this.props.markdown}
      onChange={this.props.onChange}
      plugins={this.props.plugins}/>);
  }
}

/**
 * The property types for this component
 */
ContractEditor.propTypes = {
  markdown: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  plugins: PropTypes.arrayOf(PropTypes.shape({
    onEnter: PropTypes.func,
    onKeyDown: PropTypes.func,
    renderNode: PropTypes.func.isRequired,
    toMarkdown: PropTypes.func.isRequired,
    fromMarkdown: PropTypes.func.isRequired,
    fromHTML: PropTypes.func.isRequired,
    plugin: PropTypes.string.isRequired,
    tags: PropTypes.arrayOf(PropTypes.string).isRequired,
    markdownTags: PropTypes.arrayOf(PropTypes.string).isRequired,
    schema: PropTypes.object.isRequired,
  })),
};

export { ContractEditor };
