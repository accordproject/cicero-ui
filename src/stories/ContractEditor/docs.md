# Contract Editor

#### Notes for Contract Editor

---

Supports code snippets too:

```jsx
<div>
  Foo
</div>
```