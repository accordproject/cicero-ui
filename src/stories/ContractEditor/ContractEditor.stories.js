import React, { useCallback, useState, useRef } from "react";
import { SlateTransformer } from "@accordproject/markdown-slate";
import "semantic-ui-css/semantic.min.css";
import ContractEditor from "../../ContractEditor";
import docs from "./../../ContractEditor/README.md";
import { withA11y } from "@storybook/addon-a11y";
import { withKnobs,boolean } from "@storybook/addon-knobs";
import styled from "styled-components";

const slateTransformer = new SlateTransformer();

const templateUri =
  "https://templates.accordproject.org/archives/latedeliveryandpenalty@0.15.0.cta";
const clauseText = `Late Delivery and Penalty.
----

In case of delayed delivery<if id="forceMajeure" value="%20except%20for%20Force%20Majeure%20cases%2C" whenTrue="%20except%20for%20Force%20Majeure%20cases%2C" whenFalse=""/>
<variable id="seller" value="%22Dan%22"/> (the Seller) shall pay to <variable id="buyer" value="%22Steve%22"/> (the Buyer) for every <variable id="penaltyDuration" value="2%20days"/>
of delay penalty amounting to <variable id="penaltyPercentage" value="10.5"/>% of the total value of the Equipment
whose delivery has been delayed. Any fractional part of a <variable id="fractionalPart" value="days"/> is to be
considered a full <variable id="fractionalPart" value="days"/>. The total amount of penalty shall not however,
exceed <variable id="capPercentage" value="55.0"/>% of the total value of the Equipment involved in late delivery.
If the delay is more than <variable id="termination" value="15%20days"/>, the Buyer is entitled to terminate this Contract.`;

const getContractSlateVal = () => {
  const Clause = `\`\`\` <clause src="${templateUri}" clauseid="123">
${clauseText}
\`\`\`
`;

  const defaultContractMarkdown = `# Heading One
This is text. This is *italic* text. This is **bold** text. This is a [link](https://clause.io). This is \`inline code\`.  
${Clause}
`;

  return slateTransformer.fromMarkdown(defaultContractMarkdown);
};

const Wrapper = styled.div`
  border-radius: 3px;
  border: 1px solid gray;
  margin: 50px;
  padding: 20px;
  width: min-content;
  blockquote {
    width: 80%;
    margin: 10px auto;
    padding: 1.0em 10px 1.2em 15px;
    border-left: 3px solid #484848;
    line-height: 1.4285em;
    position: relative;
  }
`;

export default { title: "'Components/Contract Editor" };

export const contractEditor = () => {
  const refUse = useRef(null);
  const [templateObj, setTemplateObj] = useState({});
  let [lockText,setLockText] = useState(true)
  let readOnly = boolean("readOnly",false);
  const [slateValue, setSlateValue] = useState(() => {
    const slate = getContractSlateVal();
    return slate.document.children;
  });

  const onContractChange = useCallback(value => {
    setSlateValue(value);
  }, []);

  const parseClauseFunction = () => console.log("parseClauseFunction");
  const loadTemplateObjectFunction = () =>
    console.log("loadTemplateObjectFunction");
  const pasteToContractFunction = () => console.log("pasteToContractFunction");

  const clausePropsObject = {
    CLAUSE_DELETE_FUNCTION: () => console.log("CLAUSE_DELETE_FUNCTION"),
    CLAUSE_EDIT_FUNCTION: () => console.log("CLAUSE_EDIT_FUNCTION"),
    CLAUSE_TEST_FUNCTION: () => console.log("CLAUSE_TEST_FUNCTION")
  };

  return (
    <Wrapper>
      <ContractEditor
        value={slateValue}
        onChange={onContractChange}
        lockText={lockText}
        readOnly={readOnly}
        ref={refUse}
        clauseProps={clausePropsObject}
        loadTemplateObject={loadTemplateObjectFunction}
        pasteToContract={pasteToContractFunction}
        onClauseUpdated={parseClauseFunction}
      />
    </Wrapper>
  );
};

contractEditor.story = {
  component: contractEditor,
  decorators: [withA11y, withKnobs],
  parameters: {
    notes: docs
  }
};
