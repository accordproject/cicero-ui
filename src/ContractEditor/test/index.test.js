import React from 'react';
import { render } from '@testing-library/react';
import { SlateTransformer } from '@accordproject/markdown-slate';
import ContractEditor from '../index';

const slateTransformer = new SlateTransformer();

const defaultContractMarkdown = `# Heading One
  This is text. This is *italic* text. This is **bold** text. This is a [link](https://clause.io). This is \`inline code\`. Fin.`;
const slate = slateTransformer.fromMarkdown(defaultContractMarkdown);
const slateFinal = slate.document.children;


const props = {
  value: slateFinal,
  onChange: () => 1,
  onClauseUpdated: () => 1,
  lockText: true,
  readOnly: true,
};

const setup = () => {
  const ref = React.createRef();
  return render(<ContractEditor {...props} ref={ref} />);
};

beforeAll(() => {
  window.getSelection = () => ({
    removeAllRanges: () => {}
  });

  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });
});

describe('<ContractEditor />', () => {
  describe('on initialization', () => {
    it('renders page correctly', () => {
      const { baseElement } = setup();
      expect(baseElement).toMatchSnapshot();
    });
  });
});
