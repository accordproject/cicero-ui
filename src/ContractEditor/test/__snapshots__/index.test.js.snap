// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContractEditor /> on initialization renders page correctly 1`] = `
<body>
  <div>
    <div
      class="ap-rich-text-editor"
      data-gramm="false"
      data-slate-editor="true"
      data-slate-node="value"
      spellcheck="true"
      style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
    >
      <h1
        class="sc-htpNat kiiZDr"
        data-slate-node="element"
        id="Heading-One"
      >
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
              Heading One
            </span>
          </span>
        </span>
      </h1>
      <p>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
              This is text. This is 
            </span>
          </span>
        </span>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <em>
              <span
                data-slate-string="true"
              >
                italic
              </span>
            </em>
          </span>
        </span>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
               text. This is 
            </span>
          </span>
        </span>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <strong>
              <span
                data-slate-string="true"
              >
                bold
              </span>
            </strong>
          </span>
        </span>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
               text. This is a 
            </span>
          </span>
        </span>
        <a
          href="https://clause.io"
        >
          <span
            data-slate-node="text"
          >
            <span
              data-slate-leaf="true"
            >
              <span
                data-slate-string="true"
              >
                link
              </span>
            </span>
          </span>
        </a>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
              . This is 
            </span>
          </span>
        </span>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <code>
              <span
                data-slate-string="true"
              >
                inline code
              </span>
            </code>
          </span>
        </span>
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
              . Fin.
            </span>
          </span>
        </span>
      </p>
    </div>
  </div>
</body>
`;
